h/views/activity.py

```diff
@@ -257,9 +255,15 @@ def toggle_tag_facet(self):
 
     def _check_access_permissions(self):
         if not self.request.has_permission('read', self.group):
-            if self.request.has_permission('join', self.group):
+            show_join_page = self.request.has_permission('join', self.group)
+            if not self.request.user:
+                # Show a page which will prompt the user to login to join.
+                show_join_page = True
+
+            if show_join_page:
```

Comment:
Ok, so the reason why visiting a private group's page when logged in but not a member of the group still works is that you added explicit logic for it here.

The comment `# Show a page which will prompt the user to login to join` may be a bit misleading because it looks like `join.html.jinja2` actually will _either_ prompt you to login (if you aren't) _or_ to join (if you're logged in).

I think the way this worked originally was that we had separate controllers for the group login page (the one that prompts you to login in order to join), the group join page, and the group search page (for members). We then used [Pyramid view configuration predicate arguments](https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/viewconfig.html?highlight=effective_principals#predicate-arguments) so that Pyramid routing would direct the request to the correct page based on whether the user was logged in or not and whether they were a member of the group or not. I thought that was nice because I kept the login to join, the join, and the search page views and templates separate from each other, and kept the logic about which page to show out of these view functions and in Pyramid routing config instead. That all seems to have been changed though, since I last looked at this. Never mind :\